### DISCUSSION TITLE
getcallstack when breakpoint is hit

### DISCUSSION THREAD

**Discussion by weidehai** [x64dbg/x64dbg#3403] [Q&A]
### Feature type

Plug-ins

### Describe the feature

is there has any plugin to getcallstack when breakpoint is hit? if not, can someone provide some ideas on how to implement it? I have read the developer documentation but couldn't find a directly usable API.

**Comment by mrexodia** [comment]
You can get the call stack of the current thread from a plugin with the following code:

```cpp
DBGCALLSTACK callstack = {};
if(!DbgFunctions()->GetCallStackEx(&callstack, /*cache=*/ false)) {
    dputs("failed to get call stack");
    return false;
}
for(int i = 0; i < callstack.total; i++) {
    const auto& e= callstack.entries[i];
    dprintf("callstack[%d] = { .addr = %p, .from = %p, .to = %p, comment = %s }", i, e.addr, e.from, e.to, e.comment);
}
BridgeFree(callstack.entries);
```

To do this at every breakpoint you can register a CBBREAKPOINT callback, but most likely you will want to create your own command with `_plugin_registercommand` and then use that command on the breakpoints you're interested in.

Could you explain what you end goal is with this? Because there might be an easier way to achieve it...

**Reply by weidehai** [reply]
i got a memory address, and set a hardware access breakpoint on it. but the breakpoint was hit many time, If I analyze when each time the breakpoint was hit to find the desired call stack, it is very time-consuming and a clumsy method. so i want to write a plugin to find call stack at each breakpoint hit,and then set breakpoint on every function call in call stack,then filter out breakpoint that hit count is 1(because i known the call stack which i want only call once)

**Comment by torusrxxx** [comment]
If you don't want to write a plugin, you can also use [printstack](https://help.x64dbg.com/en/latest/commands/script/printstack.html) command to print the call stack in log.
